pkg/util/archive.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path"
+)
+
+// ArchiveDirectory takes a directory and produces a tar ball of the contents.  If 'compress' is true, it is
+// also gzip'd.
+func ArchiveDirectory(dir string, compress bool) ([]byte, error) {
```

Comment:
Returning []byte is horrifyingly bad :)

Accept writer or return a pipe and spawn your own goroutine :)